File: atcoder.jp/PyPy3 (7.3.0)/abc265/abc265_d.py
N, P, Q, R = map(int,input().split())
A = list(map(int,input().split()))

def accumulate(l):
    res = [0] * (len(l) + 1)
    for i, n in enumerate(l):
        res[i+1] = res[i] + n
    return res

A_accum = accumulate(A)

def binary_search(ok,ng,func):
    while abs(ng - ok) > 1:
        mid = (ok+ng) // 2
        if func(mid):
            ok = mid
        else:
            ng = mid
    return ok

for x in range(N):
    y = binary_search(x, N+1, lambda m: A_accum[m] - A_accum[x] <= P)
    if A_accum[y] - A_accum[x] == P:
        z = binary_search(y, N+1, lambda m: A_accum[m] - A_accum[y] <= Q)
        if A_accum[z] - A_accum[y] == Q:
            w = binary_search(z, N+1, lambda m: A_accum[m] - A_accum[z] <= R)
            if A_accum[w] - A_accum[z] == R:
                print('Yes')
                exit()

print('No')

File: atcoder.jp/C++ (GCC 9.2.1)/abc265/abc265_d.cpp
#include <iostream>
#include <vector>
using namespace std;

template<typename T>
T binary_search(T ok, T ng, auto key)
{
    while (abs(ok - ng) > 1)
    {
        T mid = (ok + ng) / 2;
        if (key(mid)) {
            ok = mid;
        } else {
            ng = mid;
        }
    }
    return ok;
}

int main()
{
    int N;
    long long P, Q, R;
    cin >> N >> P >> Q >> R;
    vector <long long> A(N+1, 0);
    for (int i = 0; i < N; i++) {
        cin >> A.at(i+1);
        A.at(i+1) += A.at(i);
    }

    for (int x = 0; x < N; x++) {
        int y = binary_search(x, N+1, [&](int m)->bool{return A.at(m) - A.at(x) <= P;});
        if (A.at(y) - A.at(x) == P) {
            int z = binary_search(y, N+1, [&](int m)->bool{return A.at(m) - A.at(y) <= Q;});
            if (A.at(z) - A.at(y) == Q) {
                int w = binary_search(z, N+1, [&](int m)->bool{return A.at(m) - A.at(z) <= R;});
                if (A.at(w) - A.at(z) == R) {
                    cout << "Yes" << endl;
                    return 0;
                }
            }
        }
    }

    cout << "No" << endl;

    return 0;
}